### ISSUE TITLE
`useless_format` misses `format!` inside block-wrapping macros

### ISSUE THREAD

**Issue by shulaoda** [rust-lang/rust-clippy#17059]
### Description

`useless_format` silently fails to fire on `format!("{}", s)` whenever the call is the **tail expression of a block produced by another macro's expansion**. This affects rustc's entire `define_helper!`-generated family (`with_forced_trimmed_paths!`, `with_no_trimmed_paths!`, `with_no_queries!`, `with_types_for_suggestion!`, …), all of which expand to `{ let _guard = …; $e }`.

Spotted during review of [#17058](https://github.com/rust-lang/rust-clippy/pull/17058#issuecomment-4522478021):

> Even as a macro argument, using `format!("{}", some_string)` should trigger it.

### Reproducer

```rust
#![feature(decl_macro)]
#![warn(clippy::useless_format)]

// (A) macro_rules! whose body is just `$e`
macro_rules! plain_mr { ($e:expr) => { $e }; }
// (B) macro_rules! whose body wraps `$e` in a block
macro_rules! block_mr { ($e:expr) => { { let _g: i32 = 0; $e } }; }
// (C) decl_macro, body is just `$e`
pub macro plain_dm($e:expr) { $e }
// (D) decl_macro, body wraps in a block — the real `with_forced_trimmed_paths!` shape
pub macro block_dm($e:expr) { { let _g: i32 = 0; $e } }

fn s() -> String { String::from("x") }

fn main() {
    let _ = format!("{}", s());                  // (1) bare        — lints
    let _ = plain_mr!(format!("{}", s()));       // (2) mr, no block — lints
    let _ = block_mr!(format!("{}", s()));       // (3) mr + block   — MISS (bug)
    let _ = plain_dm!(format!("{}", s()));       // (4) dm, no block — lints
    let _ = block_dm!(format!("{}", s()));       // (5) dm + block   — MISS (bug)
}
```

Run with `RUSTC_BOOTSTRAP=1 clippy-driver --edition 2021 --crate-type bin -W clippy::useless_format repro.rs` (the `RUSTC_BOOTSTRAP` is only needed because of the `decl_macro` feature gate; the same miss reproduces on stable using only `macro_rules!`).

| Case | Lints? |
|------|--------|
| (1) bare                          | ✓ |
| (2) `macro_rules`, body = `$e`    | ✓ |
| (3) `macro_rules`, block-wrapped  | ✗ |
| (4) decl_macro, body = `$e`       | ✓ |
| (5) decl_macro, block-wrapped     | ✗ |

The discriminator is **block-wrapping**, not `macro_rules` vs `decl_macro`.

### Real-world instance

`clippy_lints/src/methods/unnecessary_literal_unwrap.rs` (pre-#17058):

```rust
with_forced_trimmed_paths!(format!("{}", cx.tcx.def_path_str(def.did())))
```

`def_path_str` already returns `String`, so this is the textbook `useless_format` pattern. It slipped through review precisely because the lint never saw it. `with_forced_trimmed_paths!` is defined via `define_helper!` in `compiler/rustc_middle/src/ty/print/pretty.rs`:

```rust
pub macro $name($e:expr) {
    {
        let _guard = $helper::new();
        $e
    }
}
```

i.e. case (5) — a decl_macro whose body wraps the user-supplied expression in a block.

### Root cause

`clippy_lints/src/format.rs`:

```rust
if let Some(macro_call) = root_macro_call_first_node(cx, expr)
    && cx.tcx.is_diagnostic_item(sym::format_macro, macro_call.def_id)
```

`root_macro_call_first_node` (`clippy_utils/src/macros.rs`) bails unless `first_node_in_macro(cx, expr) == Some(ExpnId::root())`. `first_node_in_macro` walks one HIR parent up; it skips a `Stmt` parent but **not** a `Block` parent.

For `block_dm!(format!(...))`: `format!`'s HIR parent is the `Block` introduced by `block_dm`'s expansion. That `Block`'s span is in `block_dm`'s expansion (sibling expansion to `format!`'s). `first_node_in_macro` returns `Some(block_dm.expn)` instead of `Some(root)`. `root_macro_call_first_node` returns `None` and the lint silently bails.

For `plain_mr!(format!(...))` the macro body is just `$e`, so `format!`'s HIR parent is the user's `let` / fn body — no macro-generated wrapper, `first_node_in_macro` returns `Some(root)`, and the lint proceeds normally.

### Version

```text
rustc 1.97.0-nightly (8b03437a8 2026-05-12)
binary: rustc
commit-hash: 8b03437a8ffc8f8b01e62ef5fce82a37ada09b12
commit-date: 2026-05-12
host: aarch64-apple-darwin
release: 1.97.0-nightly
LLVM version: 22.1.4
```

**Comment by CommanderStorm** [comment]
👋🏻 triage here

@shulaoda I believe you wanted to mark this as `Resolves ISSUE` in
- https://github.com/rust-lang/rust-clippy/pull/17060

or?
